File: pattern/1.PatternSlidingWindow/c3.js
function findLongestSubstring(str, k) {
  const count_chars = new Map();
  const length = str.length;
  let longest_str = 0;
  let start = 0;

  for (let end = 0; end < length; end++) {
    const elm = str.at(end);
    count_chars.set(elm, (count_chars.get(elm) || 0) + 1);

    while (count_chars.size > k) {
      const elm_start = str.at(start);
      count_chars.set(elm_start, count_chars.get(elm_start) - 1);
      if (count_chars.get(elm_start) === 0) {
        count_chars.delete(elm_start);
      }
      start++;
    }

    longest_str = Math.max(longest_str, end - start + 1);
  }

  return longest_str;
}

console.log(findLongestSubstring('araaci', 2));
console.log(findLongestSubstring('araaci', 1));
console.log(findLongestSubstring('cbbebi', 3));

File: pattern/1.PatternSlidingWindow/c3.py
# Longest Substring with K Distinct Characters (medium)

def find_longest_substring(s, k):
    longest = 0
    start = 0
    chars = {}

    for idx, char in enumerate(s):
        if not chars.get(char):
            chars[char] = 0
        chars[char] = chars.get(char) + 1
        while len(chars) > k and idx >= start:
            chars[s[start]] -= 1
            if chars[s[start]] == 0:
                del chars[s[start]]
            start += 1
        longest = max(longest, idx - start + 1)
    
    return longest

print(find_longest_substring("araaci",2))
print(find_longest_substring("araaci",1))
print(find_longest_substring("cbbebi",3))
